File: 2019/aoc2.py
from itertools import product


def gravity_assist(noun, verb):
    def add(a, b, write_idx):
        # print(f"writing {a} + {b} to pos {write_idx}")
        codes[write_idx] = codes[a] + codes[b]

    def multiply(a, b, write_idx):
        # print(f"writing {a} * {b} to pos {write_idx}")
        codes[write_idx] = codes[a] * codes[b]

    def terminate(_, __, ___):
        pass

    opcodes = {1: add, 2: multiply, 99: terminate}

    with open("data/2.txt") as f:
        codes = [int(c) for c in f.read().strip("\n").split(",")]
        codes[1], codes[2] = noun, verb

    i = 0
    while True:
        # print(codes[i])
        op = opcodes[codes[i]]
        if op == terminate:
            return codes[0]
        op(codes[i + 1], codes[i + 2], codes[i + 3])
        i += 4


def hunt(target):
    for n, v in product(range(100), range(100)):
        if gravity_assist(n, v) == target:
            return n, v
    raise StopIteration


def part_1():
    return gravity_assist(12, 2)


def part_2():
    n, v = hunt(19690720)
    return (100 * n) + v


if __name__ == "__main__":
    print(part_1())
    print(part_2())

File: 2019/oxidised/src/aoc2.rs
///
/// Advent of Code day 2: 1202 Error
///
/*
 * The first Intcode-based problem.
 * This appears to be the only one where the program is relevant,
 * so the solution must own the program as a primitive throughout.
 */
use crate::intcode::{Computer, From, Program};
use itertools::Itertools;

pub fn get_input() -> Vec<isize> {
    include_str!("../../data/2.txt")
        .trim()
        .split(',')
        .map(|l| l.parse::<isize>().unwrap())
        .collect()
}

pub fn part_1(program: &[isize]) -> usize {
    let mut program = program.to_owned();
    program[1] = 12;
    program[2] = 2;
    let mut program = Program::from(&program[..]);
    let mut computer = Computer::new();
    computer.run(&mut program, None).unwrap();
    program.intcodes[0] as usize
}

pub fn part_2(program: &[isize]) -> usize {
    let target = 19690720;
    let range = (0..100).permutations(2);
    for r in range {
        let (noun, verb) = (r[0], r[1]);
        let mut program = program.to_owned();
        program[1] = noun;
        program[2] = verb;
        let mut candidate = Program::from(&program[..]);
        let mut computer = Computer::new();
        computer.run(&mut candidate, None).unwrap();
        if candidate.intcodes[0] as usize == target {
            return (noun * 100 + verb) as usize;
        }
    }
    panic!("no solution found!");
}
